File: server/devflow-api/src/main/java/io/devflow/notifications/service/NotificationService.java
package io.devflow.notifications.service;

import io.devflow.common.exception.ResourceNotFoundException;
import io.devflow.notifications.dto.NotificationDto;
import io.devflow.notifications.entity.Notification;
import io.devflow.notifications.repository.NotificationRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
public class NotificationService {

    private final NotificationRepository notificationRepository;

    public NotificationService(NotificationRepository notificationRepository) {
        this.notificationRepository = notificationRepository;
    }

    @Transactional(readOnly = true)
    public Page<NotificationDto> getNotifications(UUID userId, boolean unreadOnly, Pageable pageable) {
        Page<Notification> notifications;
        if (unreadOnly) {
            notifications = notificationRepository.findByRecipientIdAndReadFalse(userId, pageable);
        } else {
            notifications = notificationRepository.findByRecipientId(userId, pageable);
        }
        return notifications.map(this::mapToDto);
    }

    @Transactional
    public void markAsRead(UUID userId, UUID notificationId) {
        Notification notification = notificationRepository.findByIdAndRecipientId(notificationId, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Notification not found"));
        notification.markRead();
        notificationRepository.save(notification);
    }

    private NotificationDto mapToDto(Notification notification) {
        return NotificationDto.builder()
                .id(notification.getId().toString())
                .title(notification.getTitle())
                .message(notification.getMessage())
                .targetUrl(notification.getTargetUrl())
                .unread(!notification.isRead())
                .createdAt(notification.getCreatedAt())
                .build();
    }
}

File: client/src/app/core/services/notification.service.ts
import { Injectable, inject } from '@angular/core';
import { ApiService } from './api.service';
import { HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Page } from './repo.service';

export interface Notification {
  id: string;
  userId: string;
  type: string;
  title: string;
  content: string;
  referenceUrl: string;
  isRead: boolean;
  createdAt: string;
}

@Injectable({
  providedIn: 'root'
})
export class NotificationService {
  private apiService = inject(ApiService);

  getNotifications(unreadOnly = false, page = 0, size = 10): Observable<Page<Notification>> {
    const params = new HttpParams().set('unreadOnly', unreadOnly).set('page', page).set('size', size);
    return this.apiService.get<Page<Notification>>('/notifications', params);
  }

  markAsRead(notificationId: string): Observable<void> {
    return this.apiService.patch<void>(`/notifications/${notificationId}/read`, {});
  }
}
